### PR TITLE
iox-#2372: Depend on @ncurses when building with Bazel

### LINKED ISSUE #2372: Use BCR @ncurses instead of relying on host's
## Brief feature description

Would be good to use the BCR's `@ncurses` (https://registry.bazel.build/modules/ncurses) instead of relying on the host's non-hermetic installation to exist.

## Detailed information

Pro: Improved hermeticity of Bazel build
Cons: none?

### PR DESCRIPTION
**Pull request by lalten**
## Notes for Reviewer

Depend on third-party `@ncurses` when building with Bazel. This is easy with Bzlmod, but I went the hard path and made it work for legacy Workspace as well.

Testing:
```console
❯ bazel run //tools/introspection:iox-introspection-client --enable_bzlmod --noenable_workspace -- --version
INFO: Analyzed target //tools/introspection:iox-introspection-client (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //tools/introspection:iox-introspection-client up-to-date:
  bazel-bin/tools/introspection/iox-introspection-client
INFO: Elapsed time: 0.061s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: bazel-bin/tools/introspection/iox-introspection-client --version
Latest official iceoryx release version: 2.95.3
❯ bazel run //tools/introspection:iox-introspection-client --noenable_bzlmod --enable_workspace -- --version
INFO: Analyzed target //tools/introspection:iox-introspection-client (71 packages loaded, 3237 targets configured).
INFO: Found 1 target...
Target //tools/introspection:iox-introspection-client up-to-date:
  bazel-bin/tools/introspection/iox-introspection-client
INFO: Elapsed time: 8.044s, Critical Path: 2.81s
INFO: 138 processes: 8 internal, 130 linux-sandbox.
INFO: Build completed successfully, 138 total actions
INFO: Running command line: bazel-bin/tools/introspection/iox-introspection-client --version
Latest official iceoryx release version: 2.95.3
```

Note also the change from `ncurses.h` vs `curses.h`. There is some background info in https://stackoverflow.com/questions/59555474/difference-between-the-headers-ncurses-h-and-curses-h
The issue is mostly that `ncurses.h` isn't available in the current BCR ncurses.
There are no implementation differences iiuc.

## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/main/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/main/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/main/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Consider a second reviewer for complex new features or larger refactorings
- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #2372

### CODE REVIEW

**Comment by elfenpiff** [comment]
@lalten from my side it looks fine, except that the ubuntu bazel build fails.

I assigned @elBoberido to this PR since he is the one with the most bazel experience.

**Review by elBoberido — COMMENTED** [review]
Looks good. Thanks for fixing this. I think you need to add the license header to the new files. You can use `bazel run //:buildifier_lint_check` to locally check whether the CI will accept the bazel files and `bazel run //:buildifier` to fix the formatting.

File: bazel/ncurses/ncurses.BUILD
[the remark this answers is not in the document]
Comment (lalten) [reply bazel/ncurses/ncurses.BUILD:1]: Can you kindly point me at instructions or examples of how this should look like for a user contribution?

[the remark this answers is not in the document]
Comment (lalten) [reply bazel/ncurses/ncurses.BUILD:1]: added something, let me know if that's good.

[the remark this answers is not in the document]
Comment (elBoberido) [reply bazel/ncurses/ncurses.BUILD:1]: Looks good.

We had a few contributors who did not want to be named in the files and they used this copyright notice
```
Copyright 2024, Eclipse Foundation and the iceoryx contributors. All rights reserved.
```

**Review by elBoberido — COMMENTED** [review]
Looks good. You need to rebase in order to make the CI run. If it's green, this can be merged.

**Review by elBoberido — APPROVED** [review]

**Pull request merged**